### PR TITLE
Add `typing._UnionSpecialForm` and add it to `isinstance`/`issubclass`

### LINKED ISSUE #7505: (🐞) `isinstance` works with `typing.Union`s on 3.10+
```py
from typing import Union

print(isinstance(1, Union[str, int]))  # error: Argument 2 to "isinstance" has incompatible type "object"; expected "Union[type, UnionType, Tuple[Union[type, UnionType, Tuple[Any, ...]], ...]]"
```
But this outputs `True`

`Union[X, Y]` is stubbed to return `object` when it really returns a `typing._UnionGenericAlias` to `typing.Union` so that would need to be fixed first.

[playground](https://mypy-play.net/?mypy=latest&python=3.10&gist=7897cc451becb2ac7a4905a2d2aa8943)

### PR DESCRIPTION
**Pull request by KotlinIsland**
`isinstance` is currently not accepting instances of `_UnionGenericAlias` aka `typing.Union`.

This change introduces a sussy fake impostor type used as a marker for these values.

Resolves #7505

### CODE REVIEW

**Comment by AlexWaygood** [comment]
> Where is the ci?

It's been sleepy lately; looks like it's woken up now though

**Comment by AlexWaygood** [comment]
@KotlinIsland, have you tested locally to check that this solves the issue?

**Comment by KotlinIsland** [comment]
Not at all. Whats the quickest way to do this?

**Comment by AlexWaygood** [comment]
Run `mypy test_script.py --custom-typeshed-dir path_to_typeshed_with_this_patch_applied`

`test_script.py` can just contain the snippet you supplied in https://github.com/python/typeshed/issues/7505#issue-1172237412. If mypy doesn't report any errors, it works!

**Comment by KotlinIsland** [comment]
Okay, just tested it and it works fine in the negative and positive.

**Comment by AlexWaygood** [comment]
The primer output indicates that mypy has no problems with this. I'm actually a little surprised at how blasé the CI is about this change.

@erictraut/ @rchen152: would this change break pyright or pytype at all?

File: stdlib/typing.pyi [hunk 1]
```diff
@@ -491,6 +491,9 @@ class _SpecialForm:
         def __or__(self, other: Any) -> _SpecialForm: ...
         def __ror__(self, other: Any) -> _SpecialForm: ...
 
+class _UnionSpecialForm(_SpecialForm):
```
[the remark this answers is not in the document]
Comment (KotlinIsland) [reply stdlib/typing.pyi]: Does it inherit the finality?

**Comment by JelleZijlstra** [comment]
> I'll also point out that I don't understand why someone would want to use the old Union form (pre PEP 604) in an isinstance call if their code is running on Python 3.10.

I can't find a realistic use case either. Some uses of the `|` operator do produce `_UnionGenericAlias` instead of `types.UnionType`:

```
>>> type(int | list[int])
<class 'types.UnionType'>
>>> type(int | List[int])
<class 'typing._UnionGenericAlias'>
```

But `isinstance()` doesn't work with `List[int]`.

This works (which surprised me the other day):

```
>>> isinstance(1, Union[int, str])
True
```

But as Eric said, why would you do that if you can use `int | str` instead? In 3.9, where the `|` operator doesn't exist yet, `isinstance()` on Union doesn't work either.

**Comment by KotlinIsland** [comment]
Because it reflects the runtime behavior? How is that not reason enough?

> This appears to be a bug in mypy

@erictraut Could you please elaborate on this, I can't see any way that this could possibly be a bug in mypy.

> I'll also point out that I don't understand why someone would want to use the old Union form (pre PEP 604) in an `isinstance` call

It's situations like this that make me so confused at the decisions made in Python, if there is no use case for this functionality, then why was it ever added to the implementation of `isinstance` in the first place?

**Comment by erictraut** [comment]
`typing.Union` is a special form that needs to be special-cased by type checkers. Mypy special-cases it in most cases where it's appropriate to do so, but it apparently doesn't have the appropriate special-casing in place for `isinstance`. This is understandable given that support for `Union` was added only recently to the runtime. The same code type checks fine in pyright.

Think about all of the other places in the type system where unions are accepted. Typeshed stubs don't need to specify "a union can be used here" because type checkers automatically handle unions as a special case.

> if there is no use case for this functionality, then why was it ever added to the implementation of isinstance in the first place?

This was added in Python 3.10 to support the new syntax in PEP 604. At runtime, `x | y` is converted into a `UnionType`. So if the runtime wants to support `isinstance(a, x | y)`, it needs to handle `UnionType` arguments for the second parameter. This means the runtime also happens to support `isinstance(a, Union[x, y])` in Python 3.10 and newer, but that wasn't the intent or motivation for the change. The fact that it happens to support this is incidental. As I mentioned above, I don't know why anyone would use the old form of `Union` with `isinstance`. It's not compatible with Python <3.10 and there's a better syntax for Python >=3.10. If your code needs to be compatible with older versions of Python, you should use `isinstance(a, (x, y))` instead.

**Comment by AlexWaygood** [comment]
> Because it reflects the runtime behavior. How is that not reason enough?

We should reflect the runtime behaviour as closely as possible at all times (in my opinion). But we also shouldn't recklessly make changes that might break large chunks of the typing ecosystem, if they don't have a good use case.

I'm personally fine if the consensus here is just "mypy should add some more special-casing". I agree that we should generally stick to the runtime implementation as closely as possible, and this PR would take us further away from that.

**Comment by KotlinIsland** [comment]
Related: python/mypy/issues/12370
Related: python/mypy/issues/12369

**Comment by rchen152** [comment]
> The primer output indicates that mypy has no problems with this. I'm actually a little surprised at how blasé the CI is about this change.
> 
> @erictraut / @rchen152: would this change break pyright or pytype at all?

pytype should be unaffected, since it has its own copies of the builtins and typing stubs.

**Comment by AlexWaygood** [comment]
I don't think there's enough consensus to make this change, so I'm closing this. But thank you for the PR; I thought it was an interesting and creative idea.

**Pull request closed without merging**